### PR TITLE
SEAB-5116: Add notebooks entry page

### PR DESCRIPTION
**Pull request by svonworl**
**Description**
This PR adds the "public" page for an individual Notebook.  I recommend that we consider this a first draft, and merge if it functions correctly and contains design/content decisions that are plausibly in the ballpark of "reasonable".  Then, as we develop the UI further, spend more time using it, and develop more experience with notebooks, we can fine-tune as a group...

The new Notebooks page has four tabs.  The "Info", "Versions", and "Files" tabs have more-or-less the same meaning and content as their Workflow counterparts.  The new "Code" tab will contain a user-friendly representation of the main notebook file (with nicely formatted code cells and rendered markdown documentation cells) and will be implemented in https://ucsc-cgl.atlassian.net/browse/SEAB-5117.

The non-public ("my-notebooks") adaptation of this page will occur as part of https://ucsc-cgl.atlassian.net/browse/SEAB-5114.

Because this PR was already big, so as not to pile on more changes and muddle the source of any problems, I only made a couple of small refactors that use the new `EntryTypeMetadata` property.  Let's do some more later.

Just for fun, I included a crude "launch with Colab" link.  Once this PR is on qa, you should be able to make a notebook on Colab, link Colab to your github, commit to a repo, install the QA github app, add a `.dockstore.yml`, and the notebook will appear on QA dockstore (make sure to publish it via `.dockstore.yml` so you can see it).  Then, you can launch via the link, and when you make future commits to the repo from Colab, the new versions should appear on Dockstore shortly thereafter.  Pretty cool.

Some GALAXY constants changed because of the addition of notebooks and how our code creates/uses the TRS schema:  https://github.com/dockstore/dockstore/pull/5429

I used the existing "non-circled" notebook icon in the spot in the notebook page where the "circled" icon, which does not yet exist, should appear.  Although I like the current icon, I have a hard time parsing it when it's small.  And, one day my brain decided that it looks like a paper towel dispenser, and now I can't unsee.  Maybe we can consider some other options (maybe a 3d book? https://seekicon.com/free-icon/book_1), and later, when we decide on one (or stick with the current), we can create the circled icon...

**Review Instructions**
Register a notebook, making sure you publish by setting the `publish` field to `true` in `.dockstore.yml` (there's no way to publish via the ui otherwise).  Find the notebook on the `/notebooks` page, then click on the link.  See the new notebook page.  Click around and make sure it works.

**Issue**
https://ucsc-cgl.atlassian.net/browse/SEAB-5116

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that your code compiles by running `npm run build`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] If this is the first time you're submitting a PR or even if you just need a refresher, consider reviewing our [style guide](https://github.com/dockstore/dockstore/wiki/Dockstore-Frontend-Opinionated-Style-Guide#pr-checklist)
- [x] Do not bypass Angular sanitization (bypassSecurityTrustHtml, etc.), or justify why you need to do so
- [x] If displaying markdown, use the `markdown-wrapper` component, which does extra sanitization
- [x] Do not use cookies, although this may change in the future
- [x] Run `npm audit` and ensure you are not introducing new vulnerabilities
- [x] Do due diligence on new 3rd party libraries, checking for CVEs
- [x] Don't allow user-uploaded images to be served from the Dockstore domain
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.
- [x] Check whether this PR disables tests. If it legitimately needs to disable a test, create a new ticket to re-enable it in a specific milestone.

### CODE REVIEW

**Comment by denis-yuen** [comment]
> I used the existing "non-circled" notebook icon in the spot in the notebook page where the "circled" icon, which does not yet exist, should appear. Although I like the current icon, I have a hard time parsing it when it's small. And, one day my brain decided that it looks like a paper towel dispenser, and now I can't unsee.

:grin: sure, I wonder if they have stable diffusion for icons

**Review by hyunnaye — APPROVED** [review]
Tested on my local and the page looks great!

File: src/app/descriptor-languages/Jupyter.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { ExtendedDescriptorLanguageBean } from 'app/entry/extendedDescriptorLanguage';
+import { SourceFile, ToolDescriptor, Workflow } from 'app/shared/swagger';
+import { Dockstore } from '../shared/dockstore.model';
+
+const JUPYTER_DOCUMENTATION_URL = Dockstore.DOCUMENTATION_URL + '/not-yet-created.html'; // TODO point at correct page
```
Comment (coverbeck) [inline src/app/descriptor-languages/Jupyter.ts:5]: I would having the real url, even if the page doesn't exist yet. Easier to only update one thing later

File: src/app/home-page/widget/entry-box/entry-box.component.ts [hunk 2]
```diff
@@ -80,6 +80,10 @@ export class EntryBoxComponent extends Base implements OnInit {
       this.helpLink = Dockstore.DOCUMENTATION_URL + '/getting-started/getting-started-with-services.html';
       this.allEntriesLink = '/my-services/';
       this.entryTypeParam = 'SERVICES';
+    } else if (this.entryType === NewEntryType.NOTEBOOK) {
+      this.helpLink = Dockstore.DOCUMENTATION_URL + '/not-yet-created.html'; // TODO point at correct page
```
Comment (coverbeck) [inline src/app/home-page/widget/entry-box/entry-box.component.ts:84]: Suggest actual link

File: src/app/shared/entry.ts [hunk 3]
```diff
@@ -125,11 +125,12 @@ export abstract class Entry implements OnDestroy {
     }
   }
 
-  protected isAppTool(url: String): boolean {
-    if (url.includes('/containers/github.com')) {
-      return true;
-    }
-    return false;
+  protected isAppTool(url: string): boolean {
+    return url.includes('/containers/github.com');
```
Comment (coverbeck) [inline src/app/shared/entry.ts:129]: Let's hope nobody ever creates an app tool named `notebooks` :) Or it's their GitHub userid. Not worth worrying about.

File: src/app/workflow/info-tab/info-tab.component.html [hunk 4]
```diff
@@ -345,17 +350,27 @@
...
             </button>
           </li>
+          <li *ngIf="(entryType$ | async) === EntryType.Notebook">
+            <strong matTooltip="Programming language used in the notebook">Programming Language</strong>:
+            {{ workflow?.descriptorTypeSubclass }}
+          </li>
           <li>
             <div *ngIf="workflow?.mode === WorkflowType.ModeEnum.STUB" class="form-inline">
-              <Strong matTooltip="Type of descriptor language used">Descriptor Type</Strong>:
+              <Strong matTooltip="Type of descriptor language used">{{
```
Comment (coverbeck) [inline src/app/workflow/info-tab/info-tab.component.html:359]: Update tooltip message for notebook?

File: src/app/workflow/info-tab/info-tab.component.html [hunk 5]
```diff
@@ -345,17 +350,27 @@
...
+              >:
               <select class="form-control input-sm" (change)="updateDescriptorType()" [(ngModel)]="temporaryDescriptorType">
                 <option *ngFor="let descriptorLanguage of descriptorLanguages$ | async" [value]="descriptorLanguage">
                   {{ 'descriptor_type' | mapFriendlyValue: descriptorLanguage }}
                 </option>
               </select>
             </div>
             <div *ngIf="workflow?.mode !== WorkflowType.ModeEnum.STUB && workflow?.descriptorType">
-              <Strong matTooltip="Type of descriptor language used">Descriptor Type</Strong>:
+              <Strong matTooltip="Type of descriptor language used">{{
```
Comment (coverbeck) [inline src/app/workflow/info-tab/info-tab.component.html:370]: Same, update tooltip message for notebook?

File: src/app/workflow/launch-third-party/launch-third-party.component.html [hunk 6]
```diff
@@ -245,6 +246,15 @@
           </a>
         </div>
       </div>
+      <div *ngIf="workflow?.descriptorType === WorkflowModel.DescriptorTypeEnum.Jupyter">
+        <a
+          target="_blank"
```
Comment (coverbeck) [inline src/app/workflow/launch-third-party/launch-third-party.component.html:251]: Needs `rel=noopener`.

**Review by coverbeck — APPROVED** [review]

**Review by kathy-t — APPROVED** [review]

File: src/assets/svg/notebook-circle.svg
[the remark this answers is not in the document]
Comment (svonworl) [reply src/assets/svg/notebook-circle.svg:1]: Yeah, it sorta looks like one of those, too.  :)

File: src/app/workflow/launch-third-party/launch-third-party.component.html [hunk 7]
```diff
@@ -245,6 +246,15 @@
...
+          href="https://colab.research.google.com/github/{{ workflow.organization }}/{{ workflow.repository }}/blob/{{
```
[the remark this answers is not in the document]
Comment (svonworl) [reply src/app/workflow/launch-third-party/launch-third-party.component.html:253]: Yes.  The launching schemes aren't standardized, so we'll use whatever capabilities each platform provides.  Github is widely supported, fortunately.  At some point down the road, if we want tighter integration, we might try to add native Dockstore support (to read/write notebooks via TRS or successor) to one or more of the platforms (for example, a JupyterHub extension that allows direct interfacing with Dockstore).

**Pull request merged**